File: backend/app/services/autumn.py
"""
SafeVision API - Autumn Credit Service
Integrates with useautumn.com for credit-based pricing.
"""

import logging
from typing import Optional, Dict, Any

import httpx

from app.config import settings

logger = logging.getLogger("safevision.autumn")

AUTUMN_BASE_URL = "https://api.useautumn.com/v1"


class AutumnService:
    """
    Handles credit checking and usage tracking via useautumn.com.
    When disabled (no API key), all requests are allowed with unlimited credits.
    """

    def __init__(self):
        self.enabled = False
        self.client: Optional[httpx.AsyncClient] = None

    def initialize(self):
        """Initialize the Autumn client if configured."""
        if settings.autumn_enabled and settings.autumn_secret_key:
            self.enabled = True
            self.client = httpx.AsyncClient(
                base_url=AUTUMN_BASE_URL,
                headers={
                    "Authorization": f"Bearer {settings.autumn_secret_key}",
                    "Content-Type": "application/json",
                },
                timeout=10.0,
            )
            logger.info("Autumn credit service enabled")
        else:
            self.enabled = False
            logger.info("Autumn credit service disabled (no API key configured)")

    async def check_credits(self, customer_id: str, feature_id: str = "detections") -> Dict[str, Any]:
        """
        Check if a customer has remaining credits for a feature.
        Returns {"allowed": bool, "balance": int|None}
        """
        if not self.enabled:
            return {"allowed": True, "balance": None}

        try:
            response = await self.client.get(
                "/entitled",
                params={"customer_id": customer_id, "feature_id": feature_id},
            )
            response.raise_for_status()
            data = response.json()
            return {
                "allowed": data.get("allowed", False),
                "balance": data.get("balance"),
            }
        except httpx.HTTPStatusError as e:
            logger.error(f"Autumn API error checking credits: {e.response.status_code} {e.response.text}")
            # Fail open - allow the request if Autumn is down
            return {"allowed": True, "balance": None}
        except Exception as e:
            logger.error(f"Autumn connection error: {e}")
            return {"allowed": True, "balance": None}

    async def track_usage(self, customer_id: str, feature_id: str = "detections", delta: int = 1) -> bool:
        """
        Track usage (decrement credits) after a successful detection.
        Returns True if tracking succeeded.
        """
        if not self.enabled:
            return True

        try:
            response = await self.client.post(
                "/events",
                json={
                    "customer_id": customer_id,
                    "feature_id": feature_id,
                    "delta": delta,
                },
            )
            response.raise_for_status()
            return True
        except httpx.HTTPStatusError as e:
            logger.error(f"Autumn API error tracking usage: {e.response.status_code} {e.response.text}")
            return False
        except Exception as e:
            logger.error(f"Autumn connection error: {e}")
            return False

    async def get_customer_info(self, customer_id: str) -> Optional[Dict[str, Any]]:
        """
        Get customer balance and plan information.
        """
        if not self.enabled:
            return None

        try:
            response = await self.client.get(f"/customers/{customer_id}")
            response.raise_for_status()
            return response.json()
        except Exception as e:
            logger.error(f"Autumn error fetching customer: {e}")
            return None

    async def create_checkout(self, customer_id: str, product_id: str) -> Optional[str]:
        """
        Create a checkout session for a customer to purchase credits.
        Returns the Stripe checkout URL.
        """
        if not self.enabled:
            return None

        try:
            response = await self.client.post(
                "/attach",
                json={
                    "customer_id": customer_id,
                    "product_id": product_id,
                },
            )
            response.raise_for_status()
            data = response.json()
            return data.get("url")
        except Exception as e:
            logger.error(f"Autumn error creating checkout: {e}")
            return None

    async def close(self):
        """Close the HTTP client."""
        if self.client:
            await self.client.aclose()


# Singleton instance
autumn_service = AutumnService()

File: middleware/src/services/autumn.ts
/**
 * useautumn.com integration — credit checking and usage tracking.
 */

import axios, { AxiosInstance } from "axios";

const AUTUMN_BASE_URL = "https://api.useautumn.com/v1";

class AutumnService {
  private enabled: boolean;
  private client: AxiosInstance | null = null;

  constructor() {
    this.enabled = false;
  }

  initialize() {
    const key = process.env.AUTUMN_SECRET_KEY || "";
    const enabled = process.env.AUTUMN_ENABLED === "true";

    if (enabled && key) {
      this.enabled = true;
      this.client = axios.create({
        baseURL: AUTUMN_BASE_URL,
        headers: {
          Authorization: `Bearer ${key}`,
          "Content-Type": "application/json",
        },
        timeout: 10_000,
      });
      console.log("Autumn credit service enabled");
    } else {
      this.enabled = false;
      console.log("Autumn credit service disabled (no API key configured)");
    }
  }

  get isEnabled(): boolean {
    return this.enabled;
  }

  async checkCredits(
    customerId: string,
    featureId: string = "detections"
  ): Promise<{ allowed: boolean; balance: number | null }> {
    if (!this.enabled || !this.client) {
      return { allowed: true, balance: null };
    }
    try {
      const res = await this.client.get("/entitled", {
        params: { customer_id: customerId, feature_id: featureId },
      });
      return {
        allowed: res.data.allowed ?? false,
        balance: res.data.balance ?? null,
      };
    } catch (err) {
      console.error("Autumn check credits error:", err);
      return { allowed: true, balance: null };
    }
  }

  async trackUsage(
    customerId: string,
    featureId: string = "detections",
    delta: number = 1
  ): Promise<boolean> {
    if (!this.enabled || !this.client) return true;
    try {
      await this.client.post("/events", {
        customer_id: customerId,
        feature_id: featureId,
        delta,
      });
      return true;
    } catch (err) {
      console.error("Autumn track usage error:", err);
      return false;
    }
  }

  async getCustomerInfo(customerId: string): Promise<any | null> {
    if (!this.enabled || !this.client) return null;
    try {
      const res = await this.client.get(`/customers/${customerId}`);
      return res.data;
    } catch (err) {
      console.error("Autumn customer info error:", err);
      return null;
    }
  }

  async createCheckout(
    customerId: string,
    productId: string
  ): Promise<string | null> {
    if (!this.enabled || !this.client) return null;
    try {
      const res = await this.client.post("/attach", {
        customer_id: customerId,
        product_id: productId,
      });
      return res.data.url ?? null;
    } catch (err) {
      console.error("Autumn checkout error:", err);
      return null;
    }
  }
}

export const autumnService = new AutumnService();
